action/protocol/staking/candidate_statemanager.go

```diff
@@ -29,24 +31,34 @@ type (
 		GetByOwner(address.Address) *Candidate
 		GetBySelfStakingIndex(uint64) *Candidate
 		Upsert(*Candidate) error
+		DebitBucketPool(context.Context, *big.Int) error
+		CreditBucketPool(context.Context, *big.Int) error
 		Commit() error
 	}
 
 	candSM struct {
 		protocol.StateManager
 		CandidateCenter
+		*bucketPool
 	}
 )
 
 // NewCandidateStateManager returns a new CandidateStateManager instance
-func NewCandidateStateManager(sm protocol.StateManager, c CandidateCenter) (CandidateStateManager, error) {
+func NewCandidateStateManager(sm protocol.StateManager, c CandidateCenter, pool *bucketPool) (CandidateStateManager, error) {
```

Comment:
I kinda of think that csm should totally own the bucketPool. it is not common a public method has a input with private type. it's better just construct bucketPool within csm creation.

in that case, csm will have a start function. One thing i am not sure is who should call statedb wrapper's start function, is the protocol using it? or factory? @CoderZhi may have some thoughts on this

Reply:
i know what you mean, in order to do that, we need to change the current way of saving candidate center in the SR's view

currently what's saved in SR's view is the address of the in-memory candidate center object, kinda hacky but works.

we need to serialize the candidate center into bytes, plus the bucket pool's serialized bytes, and store into SR view together

then we can reconstruct the candidate center + bucket pool together by reading SR's view

Reply:
or easiest way is to move struct TotalAmount into Candidate center struct